UCBChangesToScriptPics.js

```diff
@@ -0,0 +1,486 @@
+// BlockMorph menu:
+
+BlockMorph.prototype.userMenu = function () {
+    var menu = new MenuMorph(this),
+        world = this.world(),
+        myself = this,
+        hasLine = false,
+        shiftClicked = world.currentKey === 16,
+        proc = this.activeProcess(),
+        top = this.topBlock(),
+        vNames = proc && proc.context && proc.context.outerContext ?
+                proc.context.outerContext.variables.names() : [],
+        alternatives,
+        field,
+        rcvr;
+
+    function addOption(label, toggle, test, onHint, offHint) {
+        menu.addItem(
+            [
+                test ? new SymbolMorph(
+                    'checkedBox',
+                    MorphicPreferences.menuFontSize * 0.75
+                ) : new SymbolMorph(
+                    'rectangle',
+                    MorphicPreferences.menuFontSize * 0.75
+                ),
+                localize(label)
+            ],
+            toggle,
+            test ? onHint : offHint
+        );
+    }
+
+    function renameVar() {
+        var blck = myself.fullCopy();
+        blck.addShadow();
+        new DialogBoxMorph(
+            myself,
+            myself.userSetSpec,
+            myself
+        ).prompt(
+            "Variable name",
+            myself.blockSpec,
+            world,
+            blck.doWithAlpha(1, () => blck.fullImage()), // pic
+            InputSlotMorph.prototype.getVarNamesDict.call(myself)
+        );
+    }
+
+    menu.addItem(
+        "help...",
+        'showHelp'
+    );
+
+    /* UCB Script Pic Edit:
+     *  - Incorporated new method to userMenu to integrate scriptPic's and XML code.
+     *  - Code credited to Snap! Forum user Dardoro.
+     */
+    if (this.isTemplate || !this.definition) {
```

Comment:
what's this condition about? You want this option only to be available for (top) blocks that are templates, i.e. sitting in the palette or primitives. I don't understand the first constraint. The second one (only primitives) seems to aim for preventing custom block without their definition, and that's a good idea, but here you're only looking at the script's very top block. There still could be custom blocks further down inside the script, right? Would it make sense to scan the whole script for the existence of a custom block?

Reply:
If I may ... 
Latest version was 
`if( !this.isTemplate || this.definition){//for scripts & custom blocks only`

Intentions explained in the forum [post](https://forum.snap.berkeley.edu/t/enable-image-webp-format-behind-shift-property-to-store-costumes/8083/27).

> Despite its image, the custom block should be exported from the palette, NOT from the script editor.
> During a block editing session, editor content is wrongly serialized as a script:
> 
> `<script>...`
> 
> Custom block exported from the palette has a proper form.
> 
> `<blocks app="Snap! 6, https://snap.berkeley.edu" version="1">...`

Reply:
Ah, thanks, that makes sense, yes. So, we probably should be exporting not the serialized script but the / a custom block definition from the block editor. Good point.

Reply:
Thanks @DarDoro for offering the clarification :)

p.s. sorry for not tagging you in the original PR, I wasn't sure if it was you :')

Reply:
I think it would be more appropriate to allow this to be done from either location. Though, there are complexities here. In an editor instance, we can look up the actual serialized definition and attach it to the script pic. We do need to make sure the block has been saved.